langlib/lang.value/src/main/java/org/ballerinalang/langlib/value/FromJsonWithType.java

```diff
@@ -0,0 +1,290 @@
+
+/**
+ * Extern function lang.values:fromJsonWithType.
+ *
+ * @since 2.0
+ */
+@BallerinaFunction(
+        orgName = "ballerina",
+        packageName = "lang.value", version = VALUE_VERSION,
+        functionName = "fromJsonWithType",
+        args = {
+                @Argument(name = "v", type = TypeKind.ANYDATA),
+                @Argument(name = "t", type = TypeKind.TYPEDESC)
+        },
+        returnType = {
+                @ReturnType(type = TypeKind.ANYDATA),
+                @ReturnType(type = TypeKind.ERROR)
+        },
+        isPublic = true
+)
+public class FromJsonWithType {
+    private static final String AMBIGUOUS_TARGET = "ambiguous target type";
+
+    public static Object fromJsonWithType(Strand strand, Object v, TypedescValue t) {
+        BType describingType = t.getDescribingType();
+        // typedesc<json>.constructFrom like usage
+        if (describingType.getTag() == TypeTags.TYPEDESC_TAG) {
+            return convert(((BTypedescType) t.getDescribingType()).getConstraint(), v, t, strand);
+        }
+        // json.constructFrom like usage
+        return convert(describingType, v, t, strand);
+    }
+
+    public static Object convert(BType convertType, Object inputValue) {
+        return convert(convertType, inputValue, null, null);
+    }
+
+    public static Object convert(BType convertType, Object inputValue, TypedescValue t, Strand strand) {
+        try {
+            return convert(inputValue, convertType, new ArrayList<>(), t, strand);
+        } catch (ErrorValue e) {
+            return e;
+        } catch (BallerinaException e) {
+            return createError(CONSTRUCT_FROM_CONVERSION_ERROR, e.getDetail());
+        }
+    }
+
+    private static Object convert(Object value, BType targetType, List<TypeValuePair> unresolvedValues,
+                                  TypedescValue t, Strand strand) {
+        return convert(value, targetType, unresolvedValues, false, t, strand);
+    }
+
+
+    private static Object convert(Object value, BType targetType, List<TypeValuePair> unresolvedValues,
+                                  boolean allowAmbiguity, TypedescValue t, Strand strand) {
+        if (value == null) {
+            if (targetType.isNilable()) {
+                return null;
+            }
+            return createError(CONSTRUCT_FROM_CONVERSION_ERROR,
+                    BLangExceptionHelper.getErrorMessage(RuntimeErrors.CANNOT_CONVERT_NIL, targetType));
+        }
+        List<BType> convertibleTypes;
+        convertibleTypes = TypeConverter.getConvertibleTypes(value, targetType);
+        if (convertibleTypes.isEmpty()) {
+            if (TypeChecker.getType(value).getTag() == TypeTags.STRING_TAG) {
+                try {
+                    return XMLFactory.parse(((StringValue) value).getValue());
+                } catch (Throwable e) {
+                    throw createConversionError(value, targetType, e.getMessage());
+                }
+            }
```

Comment:
Why are we trying to convert it to an XML value if there aren't any convertible types? Consider the following example,
```ballerina
import ballerina/io;

public function main() {
    string foo = "Foo";
    var v = foo.fromJsonWithType(int);
    io:println(v);
}
```
The above fails with the following panic:
```
error {ballerina/lang.typedesc}ConversionError message='string' value cannot be converted to 'int': failed to parse xml: Unexpected character 'F' (code 70) in prolog; expected '<'
 at [row,col {unknown-source}]: [1,1]
```
It says `string` value cannot be converted to `int` but then has an XML parsing error as well.

Reply:
The spec says that `fromJsonWithType` should do the inverse of the `toJson` which converts `xml` into a  string. 
So, `fromJsonWithType` should be able to convert an appropriate `string` to an `xml`. 

Changing the behaviour of `TypeConverter.getConvertibleTypes()` to support this would also work, but will affect `cloneWithType` method and future methods to be implemented (if there are)

Reply:
Yeah, it doesn't make sense to try and convert to XML if I didn't specify an XML typedesc

Reply:
Yeah. Thanks!